### PR TITLE
Check that all openj9 prerequisites are up-to date on machine

### PR DESCRIPTION
**Pull request by nikolamilijevic1**
- Added checks to make sure all openj9 prerequisites are met on the machine

Signed-off-by: Nikola Milijevic <nmilijev@uwaterloo.ca>

### CODE REVIEW

**Comment by nikolamilijevic1** [comment]
@llxia @renfeiw here is draft for the pr

File: src/org/openj9/envInfo/MachineInfo.java [hunk 1]
```diff
@@ -104,17 +109,79 @@ private String execCommands(String[] commands) {
...
+	int[] versionStr2Arr(String ver) throws NumberFormatException {
+		String[] verSplit = ver.split("\\.");
+		int[] verArr = new int[verSplit.length];
+		for (int i=0; i < verSplit.length; i++) {
+			verArr[i] = Integer.parseInt(verSplit[i]);
+		}
+		return verArr;
+	}
+	
+	boolean validateVersion(String key, String ver, String reqVer) {
```
Comment (llxia) [inline src/org/openj9/envInfo/MachineInfo.java:141]: Please try to use the full name as the parameter names. 
i.e., key -> versionName
ver-> actualVersionStr
reqVer -> requiredVersionStr

File: src/org/openj9/envInfo/MachineInfo.java [hunk 2]
```diff
@@ -104,17 +109,79 @@ private String execCommands(String[] commands) {
...
+		if (matcher.find()) {
+			return matcher.group(0);
+		} else {
+			String temp = "";
+			return temp;
+		}
+	}
+
+	int[] versionStr2Arr(String ver) throws NumberFormatException {
+		String[] verSplit = ver.split("\\.");
```
Comment (llxia) [inline src/org/openj9/envInfo/MachineInfo.java:133]: Error check - handle the case where `ver` in an empty string.

**Comment by llxia** [comment]
Please format your code and provide a Grinder link once PR is updated.

File: src/org/openj9/envInfo/MachineInfo.java [hunk 3]
```diff
@@ -104,17 +117,69 @@ private String execCommands(String[] commands) {
...
+			verArr[i] = Integer.parseInt(verSplit[i]);
+		}
+		return verArr;
+	}
+
+	private boolean validateVersion(String versionName, String actualVersionStr, String requriedVersionStr) {
+		boolean isValid = true;
+		try { 
+			int[] accVer = versionStr2Arr(actualVersionStr);
+			int[] reqVer = versionStr2Arr(requriedVersionStr);
```
Comment (renfeiw) [inline src/org/openj9/envInfo/MachineInfo.java:143]: We need to handle the case that if accVer and reqVer has different length. One way to do it is to check if accVer.length == reqVer.length. If one is shorter than the other, add 0 in the shorter array. For example if accVer=1.1 reqVer=1.1.3, then append 0 in accVer which makes it 1.1.0.

File: src/org/openj9/envInfo/MachineInfo.java [hunk 4]
```diff
@@ -104,17 +117,69 @@ private String execCommands(String[] commands) {
...
+
+
+	private void validateInfo() {
+		boolean valid = true;
+		for (Map.Entry<String, String> entry : requirements.entrySet()) {
+			String version = parseInfo(infoMap.get(entry.getKey()));
+			valid &= validateVersion(entry.getKey(), version, entry.getValue());
+		}
+		if (!valid) {
+			System.out.println("\n\n");
```
Comment (renfeiw) [inline src/org/openj9/envInfo/MachineInfo.java:166]: Do we need these two extra lines here?

File: src/org/openj9/envInfo/MachineInfo.java [hunk 5]
```diff
@@ -104,17 +117,69 @@ private String execCommands(String[] commands) {
...
 		infoMap.put("ulimit", execCommands(ULIMIT_CMD));
-		infoMap.put("Ant Version", execCommands(ANT_VERSION_CMD));
-		infoMap.put("Make Version", execCommands(MAKE_VERSION_CMD));
-		infoMap.put("Perl Version", execCommands(PERL_VERSION_CMD));
-		infoMap.put("Curl Version", execCommands(CURL_VERSION_CMD));
+		infoMap.put("antVersion", execCommands(ANT_VERSION_CMD));
+		infoMap.put("makeVersion", execCommands(MAKE_VERSION_CMD));
+		infoMap.put("perlVersion", execCommands(PERL_VERSION_CMD));
+		infoMap.put("curlVersion", execCommands(CURL_VERSION_CMD));
+
```
Comment (renfeiw) [inline src/org/openj9/envInfo/MachineInfo.java:182]: please remove this empty line

File: src/org/openj9/envInfo/MachineInfo.java [hunk 6]
```diff
@@ -101,7 +116,78 @@ private String execCommands(String[] commands) {
...
+	private ArrayList<Integer> versionStr2ArrList(String version) throws NumberFormatException {
+		String[] versionSplit = version.split("\\.");
+		ArrayList<Integer> versionArr = new ArrayList<Integer>();
+		for (int i=0; i < versionSplit.length; i++) {
+			versionArr.add(Integer.parseInt(versionSplit[i]));
+		}
+		return versionArr;
+	}
+
+	private void makeSameLength (ArrayList<Integer> list, int requiredLength){
```
Comment (renfeiw) [inline src/org/openj9/envInfo/MachineInfo.java:146]: please remove the space after method name `makeSameLength`

File: src/org/openj9/envInfo/MachineInfo.java [hunk 7]
```diff
@@ -101,7 +116,78 @@ private String execCommands(String[] commands) {
 		} catch (IOException | InterruptedException e) {
 			e.printStackTrace();
 		}
-		return rt;
+		if (rt == null){
```
Comment (renfeiw) [inline src/org/openj9/envInfo/MachineInfo.java:119]: You can have `rt = "Command could not be executed";`  in the catch clause, so you don't need the if check in the end.

**Comment by nikolamilijevic1** [comment]
@llxia @renfeiw here are the internal grinder builds
| BUILD | PLATFORM | TARGET  | BUILD_LIST |
--- | --- | --- |  --- |
| Grinder_TKG/757| ppc64_aix| floatSanityTests| functional|
| Grinder_TKG/758| x86-64_windows| floatSanityTests| functional|
| Grinder_TKG/760| x86-64_linux| floatSanityTests| functional|
| Grinder_TKG/761| x86-64_mac| floatSanityTests| functional|

File: src/org/openj9/envInfo/MachineInfo.java [hunk 8]
```diff
@@ -100,21 +115,88 @@ private String execCommands(String[] commands) {
...
+					isValid = false;
+				} else if (reqVer.get(i) < accVer.get(i)) {
+					break;
+				}
+			}
+			if (!isValid) {
+				System.out.println("Error: required " + versionName + ": " + requriedVersionStr + ". Installed version: " + actualVersionStr);
+			}
+		} catch (NumberFormatException e){
+			System.out.println("Warning: version information (" + versionName + " : " + requriedVersionStr + ") cannot be extracted.");
```
Comment (llxia) [inline src/org/openj9/envInfo/MachineInfo.java:172]: This message is misleading. The output from the Grinder:
```
Warning: version information (antVersion : 1.9.6) cannot be extracted.
```

The problem is that we cannot get the `actualVersionStr `, not `requriedVersionStr `. Please change it to:
```
System.out.println("Warning: versionName + " information cannot be extracted.");
System.out.println(versionName + " output: " + actualVersionStr);

```

Comment (nikolamilijevic1) [reply src/org/openj9/envInfo/MachineInfo.java]: updated 👍 https://hyc-runtimes-jenkins.swg-devops.com/view/Test_grinder/job/Grinder_TKG/765/consoleFull

**Pull request merged**